keps/sig-storage/1472-storage-capacity-tracking/README.md

```diff
@@ -1189,9 +1189,54 @@ based on storage capacity:
   to available storage and thus could run on a new node, the
   simulation may decide otherwise.
```

Comment:
A separate issue is that Cluster Autoscaler simulates scheduling of all the pods (binpacking) before making any autoscaling decisions. The way that capacity is defined here (and how https://github.com/kubernetes/autoscaler/pull/3887 uses the same CSIStorageClass to represent local storage of potentially hundreds of nodes) I think means that CA will not be able to calculate how many nodes are needed for any use-case where available storage is the limiting factor for scheduling.

ex. Assume I have a 1000 pending pods that have relatively low cpu/memory requests, but require all the local storage of a node. CA would think all those pods can fit on just a few nodes (on a single node if cpu/memory requests are low enough) and only add this many nodes. Admittedly, once the nodes are created some pods will be scheduled, CA will see that remaining pods are unschedulable again and trigger another (equally small) scale-up. So theoretically the situation will eventually recover. In practice the latency can be arbitrarily long, making the feature completely unusable in sufficiently large clusters.

Reply:
I understand that's how CSI specification works and we may not be able to change it, but it seems that capacity specification isn't strictly defined. We could say that for purposes of autoscaling we treat it as sum of all volumes that can be created. This is not a perfect prediction as it doesn't take fragmentation, etc into account, but it seems much more realistic than assuming we can create infinite number of volumes and the only limitation is on the size of individual volume.

We could further explicitly say that any driver where this assumption is too far off is incompatible with autoscaling. Given how popular the autoscaling is in cloud environments, this can hopefully generate some incentive for driver authors to interpret capacity that way.

Reply:
My concern is that the API is even problematic for kube-scheduler itself. On a sudden scale up, multiple pods will end up failing. then we need to wait for them to fail to create replacement pods. If you are using the Job API, you could hit the `backoffLimit` and the Job would fail. Similarly, other third party APIs (argo, airflow) are less resistant to failures.

> We could say that for purposes of autoscaling we treat it as sum of all volumes that can be created. This is not a perfect prediction as it doesn't take fragmentation, etc into account, but it seems much more realistic than assuming we can create infinite number of volumes and the only limitation is on the size of individual volume.

I agree that such interpretation would be better. Even if it's not a perfect prediction, it would at least generate less failed pods.

As the API stands today, I don't think I'm confident graduating it to GA, specially if changing the interpretation of a field is under consideration. Maybe we instead need to add a new field for the total capacity and scheduler needs to respect in combination with the existing `.status.availableCapacity`. If the field is not set, it means that the capacity is unlimited and we can assume that drivers who choose to leave it empty understand the consequences.

Do you think adding that field is possible @pohly? Do you think that's better for scheduling/autoscaling @MaciekPytel?

If you agree, we should go with a v1beta2.

Reply:
FYI @liggitt @roycaihw (API reviewers)

Reply:
> I understand that's how CSI specification works and we may not be able to change it, but it seems that capacity specification isn't strictly defined.

That's why I added a more strictly defined "maximum volume size" to the spec. That is what Kubernetes uses if available, instead of the more loosely defined "capacity".

> We could say that for purposes of autoscaling we treat it as sum of all volumes that can be created. This is not a perfect prediction as it doesn't take fragmentation, etc into account, but it seems much more realistic than assuming we can create infinite number of volumes and the only limitation is on the size of individual volume.

I don't think that making assumptions about how CSI drivers handle storage capacity is going to help. We've discussed that at length in SIG Storage and in CSI meetings and the consensus there was that there are always some vendor specific exceptions that cannot be handled in a generic way.

> On a sudden scale up, multiple pods will end up failing.  then we need to wait for them to fail to create replacement pods.

Failing how? They won't get scheduled unless their volumes were successfully created. Until then they remain in the queue and will eventually land on some node.

Reply:
Ah sorry, I was somehow thinking that the Pod would be scheduled and fail in the kubelet.

> That's why I added a more strictly defined "maximum volume size" to the spec. That is what Kubernetes uses if available, instead of the more loosely defined "capacity".

I don't see this here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1472-storage-capacity-tracking/README.md#csistoragecapacity

Can you please update?

Although it doesn't seem to add much, as that is the interpretation we are going with for the older ".status.availableCapacity". Then why do we need both?

> I don't think that making assumptions about how CSI drivers handle storage capacity is going to help. We've discussed that at length in SIG Storage and in CSI meetings and the consensus there was that there are always some vendor specific exceptions that cannot be handled in a generic way.

Right, there might be drivers for which we won't be able to do a prediction. But for some of them we should be able to do a reasonable prediction, and the API doesn't have any way to express that.

Reply:
> Can you please update?

Right, that section is no longer up-to-date. I'll update it.

> Although it doesn't seem to add much, as that is the interpretation we are going with for the older ".status.availableCapacity". Then why do we need both?

Because the CSI spec never specified in detail what `CapacityResponse.available_capacity` actually means. Retroactively updating the CSI spec to assign some new meaning would have been a breaking change, whereas adding a new field is possible. Kubernetes uses the `available_capacity` field only if `maximum_volume_size` is not available. That's then still better than not doing anything at all, even if it is less precise.

"Maximum volume size" is what the CSI community was comfortable with because it's both useful and defined precisely enough to be unambiguous. "Total capacity" was rejected because it's only useful in combination with additional assumptions (like "storage is linear").

The PR which added the field has some of that discussion: https://github.com/container-storage-interface/spec/pull/470

Reply:
Is it possible to have a "min available capacity" (either by reusing existing `available_capacity` or via a separate field to avoid backwards compatibility issues) or is it also not viable? The semantics I have in mind would be: capacity that is guaranteed to be available, regardless of individual volume sizes. Cluster Autoscaler would then be able to use it as a lower bound for binpacking. Is this too generic as well?

Reply:
I'm not sure I fully understand the intended semantic, but it smells like it has the same problem: guaranteed to be available for what? Creating two arbitrary volumes that together don't exceed that amount?

Reply:
Yes. I understand it is not possible to tell up front what's the total capacity of a set of volumes that could be created, because it depends on things like sizes and order of creating these volumes. However, there should be a lower bound to that number (the trivial, yet useless lower bound is 0). The question I don't know the answer to is whether it is always possible to effectively calculate a lower bound that is actually useful.

Reply:
So a scheduling cycle would start with "x bytes available" and each volume that gets tentatively assigned to a node  (= selected node label set) reduces that amount by the size of the volume?

It seems like a rather crude approach and there are open questions like how the scheduler should deal with PVCs that are already in the process of getting provisioned for the node, but I can imagine that this might help in some scenarios. 

I just don't know whether the extra complexity is justified. The right approach for tackling that question would be to draft a KEP that hashes out all of the design details and lists the use cases that get improved by it. Then any new fields or functionality related to that KEP can be tracked separately, starting with alpha and a new feature gate.

Reply:
I'm looking at the graduation as a gate that prevents people from dropping an API/implementation that is not perfect and not doing future work to improve the situation, because it's already GA.

Not saying that you would do that, just saying that it's a general concern.

Reply:
I understand the concern. I am motivated to continue to improve Kubernetes in this area, but so far I haven't seen enough evidence that "the scheduling latency is not acceptable" to make it a priority.

"My pod isn't getting scheduled when using all of the storage capacity" is a more obvious problem and gets solved by the code for this KEP, therefore I want to get that solution into the hands of users before tackling additional aspects.

"Perfect is the enemy of good" applies here...

Reply:
I think the fundamental issue here is a conflict in requirements:

>  "Total capacity" was rejected because it's only useful in combination with additional assumptions (like "storage is linear").

Cluster Autoscaler doesn't process pods sequentially like scheduler, it predicts the result of scheduling of all the pending pods in order to see how many nodes are needed. ie. we need to be able to predict the result of scheduling all of the pending pods before we actually schedule the first one. I don't see how it's possible if we don't make some additional assumptions.

I'd much rather we explicitly list 'incompatibility with CA' as a limitation of the current design and later seek to address this, by extending CSI spec with some optional stronger assumptions that would allow CA to support plugins meeting those assumptions (a 'total capacity' field is one an example solution, but I'm not claiming it's the best possible one).

Reply:
> In terms of API stability, we cannot change the interpretation of existing fields. Any new semantic we want in the future would have to be a new field.

Not saying it's the way to go, but a new apiversion *could* repurpose a field name as long as the round-trip logic converts, right?

Reply:
I don't think we need to go that far. Both existing fields are intentionally marked as optional. We can always add new ones and stop populating the old ones, if we find that this will work better. Nothing in the current discussion indicates that there'll be a need for that, though, therefore I am confident that this API is good for GA.

Reply:
I had an idea how we can support simulated volume provisioning in the autoscaler without (at least initially) extending this API here or the CSI spec:
- the CSI driver must report "total capacity" in `GetCapacityResponse.available_capacity` and support `maximum_volume_size`
- the CSI driver must be local to nodes and have a single topology key with the Kubernetes node name as value
- the CSIDriver object must have a special annotation for the autoscaler ("linear, local storage")
- nothing changes for kube-scheduler
- autoscaler simulates scheduling like this:
  - when cloning an existing node to create a node group, it replaces the node name in the CSI driver's topology label
  - it clones the CSIStorageCapacity objects for each storage class, changes their topology and removes the `MaximumVolumeSize` field: what remains is the total capacity without any volumes allocated
  - when the "selected node" label gets set for a PVC while scheduling a pod, the "AvailableCapacity" gets reduces by the size of that PVC: this simulates volume creation on that node and ensures that eventually pods no longer fit the node

There are some implementation challenges, for example the volume binder code that sets "selected node" might not get called at the moment by autoscaler. The in-tree volume binder code must be made a bit more flexible to accept stores instead of creating informers itself, but other than that it doesn't need any special logic. It's also not a generic solution, so we must have a good understanding of which CSI drivers are going to use this to check in advance that they fit the assumptions.

Reply:
If `MaximumVolumeSize` is simply removed, Cluster Autoscaler may decide to provision a node even though a workload can't get scheduled there due to its request exceeding the max size (but being less than total capacity). Proper CA support needs to somehow get this information.

Also, this doesn't really help in scale-from-0 scenario. CA needs to somehow know how new nodes will look like, without an existing sample node to clone.

Reply:
I probably shouldn't have posted that idea because we are now back at discussing a solution that doesn't belong into this KEP :sweat_smile: Let's continue anyway, but let's beware that this isn't blocking merging of this PR.

The entire approach of modeling volume provisioning in the autoscaler has to make assumptions about the storage system. The key assumption here is "storage is linear" in which case "maximum volume size" becomes irrelevant because any volume whose size is smaller than the remaining capacity can get created. If that isn't close enough to reality, a more complex model will be needed which then might include some information about "volumes have a minimum size, a maximum size, size must be aligned with X GB, etc.".

> Also, this doesn't really help in scale-from-0 scenario. CA needs to somehow know how new nodes will look like, without an existing sample node to clone.

The CSI driver for local storage cannot help with scale-from-0 because it isn't running. The cloud provider or some other central component would have to be extended to provide information for such a scenario. That's a different approach that gives up flexibility (cannot use arbitrary CSI drivers) in favor of scale-from-0 support.

Reply:
Sorry for starting a solution discussion here, happy to take it elsewhere :slightly_smiling_face: 

> The entire approach of modeling volume provisioning in the autoscaler has to make assumptions about the storage system. The key assumption here is "storage is linear" in which case "maximum volume size" becomes irrelevant because any volume whose size is smaller than the remaining capacity can get created. If that isn't close enough to reality, a more complex model will be needed which then might include some information about "volumes have a minimum size, a maximum size, size must be aligned with X GB, etc.".

Yes, if we want proper autoscaling support, CA has to use a more complex model, which will make sense for most/all storage implementations out there.

> The CSI driver for local storage cannot help with scale-from-0 because it isn't running. The cloud provider or some other central component would have to be extended to provide information for such a scenario. That's a different approach that gives up flexibility (cannot use arbitrary CSI drivers) in favor of scale-from-0 support.

I think this could work through a separate API: different driver extensions would create different objects for CA to use for resource predictions.